### PR TITLE
Issue 1123: EPP cache not working

### LINKED ISSUE #1123: EPP cache not working
### Current Behavior

The cache is empty when accessing a property getter resulting in more epp calls when in any of the manage domain pages.

### Expected Behavior

When using a domain property getter, it should pull from cache just as it does via command line or in the unit-tests. 

### Steps to Reproduce

1. add a comment on the __init__ and __get__ functions found in domain.py. May also be helpful to add a comment in the get property function
1. go to the domain management main page 
2. see the init call being called twice

### Environment

_No response_

### Additional Context

When going to the manage domain pages, this causes the domain object to be reintialized twice. Each initialization is being called before each getter call, once for the domain.nameserver call and once for domain.security email call. This should not be happening  and is not typical of Django; as normally, an app instance should stay in the context you’re in while using the app. This happens each time we do a domain.[property-value] no the security email page and the nameserver page. As a result we are seeing double the calls needed and resolving this would cut call count by half. 

[see enghuddle for 10/2](https://docs.google.com/document/d/16S6VNKOYW2HI7bKwOwOdQzon5SZA0ZgkMkGfQHLZoTw/edit#heading=h.gw2e13758zoc)
[slack thread](https://cisa-corp.slack.com/archives/C05BGB4L5NF/p1696006778478359)

### Issue Links

_No response_

### PR DESCRIPTION
**Pull request by dave-kennedy-ecs**
## Ticket

Resolves #1123 

## Changes

- Added session cache to Domain views so that domain can persist in session cache between calls
- Cleaned up code in Domain views for retrieving and setting domain to and from cache and db
- Refactored _fetch_cache in Domain model

## Context for reviewers

Cache was not persisting between each page load or form submit.  This was causing EPP calls, sometimes several, on every page load.  And the domain object was initialized with each page load or form submit.

With these changes, domain object is initialized once.  And EPP calls are made only when necessary.  Page loads should be noticeably quicker.

## Setup

No setup necessary for testing.  

Just Navigate to the Home page, and click on Manage a Domain.  Move between the different pages for managing a domain.  Change values in the forms and submit.  Check responsiveness.

#### Satisfied acceptance criteria and met development standards

- [x] Met the acceptance criteria, or will meet them in a subsequent PR
- [x] Created/modified automated tests
- [x] Added at least 2 developers as PR reviewers (only 1 will need to approve)
- [x] Messaged on Slack or in standup to notify the team that a PR is ready for review
- [ ] Changes to “how we do things” are documented in READMEs and or onboarding guide
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Original Developer)

- [x] All new functions and methods are commented using plain language
- [ ] Did dependency updates in Pipfile also get changed in requirements.txt?
- [x] Interactions with external systems are wrapped in try/except
- [x] Error handling exists for unusual or missing values

#### Validated user-facing changes (if applicable)

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)
- [ ] Add at least 1 designer as PR reviewer

#### Reviewed, tested, and left feedback about the changes

- [x] Pulled this branch locally and tested it
- [x] Reviewed this code and left comments
- [ ] Checked that all code is adequately covered by tests
- [x] Made it clear which comments need to be addressed before this work is merged
- [ ] If any model was updated to modify/add/delete columns, makemigrations was ran and the assoicated migrations file has been commited.

#### Ensured code standards are met (Code reviewer)

- [ ] All new functions and methods are commented using plain language
- [x] Interactions with external systems are wrapped in try/except
- [x] Error handling exists for unusual or missing values
- [ ] (Rarely needed) Did dependency updates in Pipfile also get changed in requirements.txt?

#### Validated user-facing changes as a developer

- [ ] New pages have been added to .pa11yci file so that they will be tested with our automated accessibility testing
- [x] Checked keyboard navigability
- [ ] Meets all designs and user flows provided by design/product
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers, the suggestion is to use ones that the developer didn't (check off which ones were used)
  - [x] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

**Note:** Multiple code reviewers can share the checklists above, a second reviewers should not make a duplicate checklist

#### Verified that the changes match the design intention

- [ ] Checked that the design translated visually
- [ ] Checked behavior
- [ ] Checked different states (empty, one, some, error)
- [ ] Checked for landmarks, page heading structure, and links
- [ ] Tried to break the intended flow

#### Validated user-facing changes as a designer

- [ ] Checked keyboard navigability
- [ ] Tested general usability, landmarks, page header structure, and links with a screen reader (such as Voiceover or ANDI)

- [ ] Tested with multiple browsers (check off which ones were used)
  - [ ] Chrome
  - [ ] Microsoft Edge
  - [ ] FireFox
  - [ ] Safari

- [ ] (Rarely needed) Tested as both an analyst and applicant user

### CODE REVIEW

File: src/registrar/views/domain.py
Comment (zandercymatics) [inline src/registrar/views/domain.py:82]: Love this implementation

**Review by zandercymatics — APPROVED** [review]
Looks great! This change with the connection pooling in my opinion I think completely fixes the performance issue. Just a few small cleanup items.

I did find one bug though with security email: if you open a new domain, save a contact then save it again, and delete the PublicContact object through /admin, it will throw a 500 error, a ContactError of type "No contact was found in cache or the registry".

This happens because the getter for contact expects there to be a DB object (a PublicContact) after a contact is added to cache. The cache stores registry ids for each PublicContact, then it just pairs the associated Contact with the id. 

This is something I think can be handled as an edge case within the get_contact_in_keys() function, wherein: 
if desired.count() == 0:
    (logic in the for loop for _fetch_contacts())
    
Approving for now as I am not sure if this is something to handle in this ticket or another ticket. I'd recommend waiting for another engineer to take a look at this as well before merging for that reason, though

File: src/registrar/models/domain.py
Comment (abroddrick) [reply src/registrar/models/domain.py:1703]: logger.debug will not appear on sandboxes I believe. We had that set to info level and above, so we will effectively never see this if it stays debug (unless a setting has been changed)

**Comment by dave-kennedy-ecs** [comment]
> I did find one bug though with security email: if you open a new domain, save a contact then save it again, and delete the PublicContact object through /admin, it will throw a 500 error, a ContactError of type "No contact was found in cache or the registry".
> 
> This happens because the getter for contact expects there to be a DB object (a PublicContact) after a contact is added to cache. The cache stores registry ids for each PublicContact, then it just pairs the associated Contact with the id.

I am fairly certain that Rachid's PR on 1103 fixes exactly this.  So, when both this PR and that PR are merged, we should test the scenario you are bringing up.

File: src/registrar/views/domain.py
Comment (dave-kennedy-ecs) [reply src/registrar/views/domain.py:82]: thanks

File: src/registrar/models/domain.py
Comment (dave-kennedy-ecs) [reply src/registrar/models/domain.py:1733]: good catch.  i removed the redundancy

File: src/registrar/models/domain.py
Comment (dave-kennedy-ecs) [reply src/registrar/models/domain.py:1703]: good to know about the log levels in sandbox and local, thanks

**Pull request merged**
